jerry-core/ecma/base/ecma-module.c

```diff
@@ -13,246 +13,864 @@
+      if (context_p->local_exports_p != NULL)
+      {
+        /* 15.2.1.16.3 / 4 */
+        JERRY_ASSERT (context_p->local_exports_p->next_p == NULL);
+        ecma_module_names_t *export_names_p = context_p->local_exports_p->module_names_p;
+        while (export_names_p != NULL)
+        {
+          if (ecma_compare_ecma_strings (current_export_name_p, export_names_p->imex_name_p))
+          {
+            if (found)
+            {
+              /* This is an ambigous export. */
+              found_record.module_p = NULL;
+              found_record.name_p = NULL;
+              break;
+            }
+
+            /* The current module provides a direct binding for this export. */
+            found = true;
+            found_record.module_p = current_module_p;
+            found_record.name_p = export_names_p->local_name_p;
+            break;
+          }
+
+          export_names_p = export_names_p->next_p;
+        }
+      }
 
-      if (exported_name_p == NULL)
+      if (found)
       {
-        ecma_deref_ecma_string (prop_name_p);
+        /* We found a resolution for the current frame, return to the previous. */
+        ecma_module_resolve_stack_pop (&stack_p);
         continue;
       }
 
-      parser_module_names_t *new_name_p =
-      ecma_parser_module_compare_property_name_with_import (module_context_p->imports_p,
-                                                            exported_name_p);
-
-      if (new_name_p != NULL)
+      /* 15.2.1.16.3 / 5 */
+      ecma_module_node_t *indirect_export_p = context_p->indirect_exports_p;
+      while (indirect_export_p != NULL)
       {
-        ecma_property_t *new_property_p;
-        ecma_create_named_data_property (global_obj_p,
-                                         new_name_p->import_name_p,
-                                         ECMA_PROPERTY_NOT_WRITABLE,
-                                         &new_property_p);
+        ecma_module_names_t *export_names_p = indirect_export_p->module_names_p;
+        while (export_names_p != NULL)
+        {
+          if (ecma_compare_ecma_strings (current_export_name_p, export_names_p->imex_name_p))
+          {
+            /* 5.2.1.16.3 / 5.a.iv */
+            ecma_module_resolve_stack_push (&stack_p,
+                                            indirect_export_p->module_request_p,
+                                            export_names_p->local_name_p);
+          }
+
+          export_names_p = export_names_p->next_p;
+        }
+
+        indirect_export_p = indirect_export_p->next_p;
+      }
 
-        ecma_named_data_property_assign_value (global_obj_p,
-                                               ECMA_PROPERTY_VALUE_PTR (new_property_p),
-                                               prop_pair_p->values[i].value);
+      /* We need to check whether the indirect exports resolve to anything. */
+      continue;
+    }
```

Comment:
Is this indentation correct? Would be better to mark with a comment which loop ends with this right brace.

Reply:
Looks right to me, I'll add a comment nevertheless.